e2e/deployment.go

```diff
@@ -0,0 +1,173 @@
+}
+
+// deleteDeploymentApp deletes the deployment object.
+func deleteDeploymentApp(name, ns string, clientSet kubernetes.Interface, deployTimeout int) error {
+	timeout := time.Duration(deployTimeout) * time.Minute
+	err := clientSet.AppsV1().Deployments(ns).Delete(context.TODO(), name, metav1.DeleteOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to delete deployment: %w", err)
+	}
+	start := time.Now()
+	e2elog.Logf("Waiting for deployment %q to be deleted", name)
+
+	return wait.PollImmediate(poll, timeout, func() (bool, error) {
+		_, err := clientSet.AppsV1().Deployments(ns).Get(context.TODO(), name, metav1.GetOptions{})
+		if err != nil {
+			if isRetryableAPIError(err) {
+				return false, nil
+			}
+			if apierrs.IsNotFound(err) {
+				return true, nil
+			}
+			e2elog.Logf("%q deployment to be deleted (%d seconds elapsed)", name, int(time.Since(start).Seconds()))
+
+			return false, fmt.Errorf("failed to get deployment: %w", err)
+		}
+
+		return false, nil
+	})
+}
+
+// waitForDeploymentInAvailableState wait for deployment to be in Available
+// condition.
+func waitForDeploymentInAvailableState(name, ns string, clientSet kubernetes.Interface, deployTimeout int) error {
+	timeout := time.Duration(deployTimeout) * time.Minute
+	start := time.Now()
+	e2elog.Logf("Waiting up to %q to be in Available state", name)
+
+	return wait.PollImmediate(poll, timeout, func() (bool, error) {
+		d, err := clientSet.AppsV1().Deployments(ns).Get(context.TODO(), name, metav1.GetOptions{})
+		if err != nil {
+			if isRetryableAPIError(err) {
+				return false, nil
+			}
+			e2elog.Logf("%q deployment to be Available (%d seconds elapsed)", name, int(time.Since(start).Seconds()))
+
+			return false, err
+		}
+		cond := deploymentutil.GetDeploymentCondition(d.Status, appsv1.DeploymentAvailable)
+
+		return cond != nil, nil
+	})
+}
+
+// Waits for the deployment to complete.
+func waitForDeploymentComplete(name, ns string, clientSet kubernetes.Interface, deployTimeout int) error {
+	var (
+		deployment *appsv1.Deployment
+		reason     string
+		err        error
+	)
+	timeout := time.Duration(deployTimeout) * time.Minute
+	err = wait.PollImmediate(poll, timeout, func() (bool, error) {
+		deployment, err = clientSet.AppsV1().Deployments(ns).Get(context.TODO(), name, metav1.GetOptions{})
+		if err != nil {
+			if isRetryableAPIError(err) {
+				return false, nil
+			}
+			e2elog.Logf("deployment error: %v", err)
+
+			return false, err
+		}
+
+		// TODO need to check rolling update
+
+		// When the deployment status and its underlying resources reach the
+		// desired state, we're done
+		if deployment.Status.Replicas == deployment.Status.ReadyReplicas {
+			return true, nil
+		}
+		e2elog.Logf(
+			"deployment status: expected replica count %d running replica count %d",
+			deployment.Status.Replicas,
+			deployment.Status.ReadyReplicas)
+		reason = fmt.Sprintf("deployment status: %#v", deployment.Status.String())
```

Comment:
why `%#v` ?

Reply:
This function is just a move from the previous pod.go, I have not made any logic changes to it.

If someone confirms there is no meaning for `%#v` use I can remove it, at least I don't know :-)

Reply:
Doesn't look like unintentional:
https://github.com/kubernetes/kubernetes/blob/master/test/utils/deployment.go#L103